osquery/tables/system/windows/prefetch.cpp

```diff
@@ -0,0 +1,307 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include "osquery/filesystem/fileops.h"
+#include <osquery/core/core.h>
+#include <osquery/core/tables.h>
+#include <osquery/filesystem/filesystem.h>
+#include <osquery/logger/logger.h>
+#include <osquery/sql/dynamic_table_row.h>
+#include <osquery/tables/system/windows/prefetch.h>
+#include <osquery/utils/conversions/join.h>
+#include <osquery/utils/conversions/windows/windows_time.h>
+#include <osquery/utils/windows/lzxpress.h>
+
+#include <boost/filesystem.hpp>
+#include <boost/format.hpp>
+
+namespace osquery {
+namespace tables {
+
+const std::string kPrefetchLocation = (getSystemRoot() / "Prefetch\\").string();
+
+// Convert End of string UCHAR vectors to string
+std::string ucharToString(const std::vector<UCHAR>& data) {
+  int data_char = 0;
+  std::string data_string;
+  while (data[data_char] != '\0') {
```

Comment:
So just to confirm it looks like the code above is passing the a string_view instead of a vector and is calculating the final string base on the string_view/data size?

Just a few clarifications:
so according to prefetch format [docs (libscca)](https://github.com/libyal/libscca/blob/main/documentation/Windows%20Prefetch%20File%20(PF)%20format.asciidoc) the filename, access files/directories should/need to have null terminated/end of string characters.  
The way accessed data is parsed/handled in prefetch files (base on my understanding) is that the prefetch file contains 3 important elements:

- the offset to the start of the access data block/array 
- the size of the whole data block/array (the size of all the accessed files/directories)
- the array/block of accessed data (data meaning accessed files and accessed directories  
\* data = accessed files, accessed directories

The sizes of the individual filenames and directories are not known.  So what ucharToString is doing is looping through each filename/directory in the data block/array until it reaches the end string/null char.  
it then returns the string and calculates the size, and then erases the size from the data block/array and keeps going until the data block/array is zero.

Base on my understanding the code above would calculate the string size as the whole data block/array instead of the individual files/directories? Since there is no way to calculate the individual filenames or directories?

Reply:
The real size of the string is calculated by `strnlen`, which still scans for the null terminator, but limits the scan to the size of the `string_view`.
Then the string is constructed using the `string_view` pointer and the found string size, which doesn't result in any additional scan.

Additionally `string_view` is there to avoid forcing the caller to make a copy of a vector to limit the range of data you want to search the null terminator in.
There could've been other methods, like passing an index and a limit size, or as many C++ algorithm do, accept a begin and end iterator, where you could've used `std::distance` on to find the size of the range passed.

Reply:
oops, yeah i see that now. That makes sense.  
So i was able to add `strnlen` but encountered issues getting the length.  Since the filenames are UTF16 encoded `strnlen` is only getting a length of 1?  
ex: `37 00 5A 00 46 00 4D 00 2E 00 45 00 58 00 45 00 00 00`
I added a loop to continue getting the characters until the null terminated string is found `0000`.  I tried researching for a better way, not sure if there better ways to make `strnlen` work with UTF16 strings?

Reply:
My bad! I didn't realize they were UTF16 strings.
Searching for `\0\0` though isn't necessarily correct though, right? It could take a null byte from the previous character incorrectly.

You need to treat them as `std::wstring` or wchar_t*, then we can properly take their size.
Admittedly though I'm not super familiar with the whole `char *` to `wchar_t *` (or `std::wstring`) dance.
@farfella can I ask your insight?

I recall this commit added some support to convert to/from UTF16: https://github.com/osquery/osquery/commit/60172a37743ebb58abc9ff06b906fbc188e6ae92
But it requires you to have a `std::string`, which is probably returned by an API, not from file. For a string coming from a file I believe we also have to do deserialization because it depends on how it has been saved (little endian or not), right?

Moreover, it would be ideal to avoid unnecessary copies as much as possible (due to the fact that we start from a `std::vector<UCHAR>`) .

Given that these strings are read from file to be then passed to sql as results, we don't necessarily need to pass through a `std::wstring`.
Could we add a helper that accepts a `char *` and a `size` and deserializes the UTF16 string, converting it directly to an UTF8 `std::string`?

Reply:
so currently the `ucharToString` function is looking at every other char in the string_view.
ex: `37 00 5A 00 46 00 4D 00 2E 00 45 00 58 00 45 00 00 00` (Which is the string `7ZFM.EXE` ) it starts at [0] then [2],[4],[6], etc. Until it reaches a null byte. It shouldnt encounter any issues with null bytes from previous characters

Reply:
Hmmm. Let's think about this from a different angle. Based on the libscca documentation link above, can these be represented as structures and read in as such? So, in this case, we have two structures of sorts:
```
#pragma pack(push, 1)
typedef struct _PREFETCH_COMPRESSED_HEADER {
    DWORD Signature;
    DWORD TotalUncompressedSize;
    BYTE CompressedData[1]; // arbitrary size
} PREFETCH_COMPRESSED_HEADER;

typedef struct _PREFETCH_FILE_HEADER {
    DWORD Version;
    DWORD Signature;
    DWORD Reserved1;
    DWORD FileSize;
    WCHAR FileName[30];
    DWORD Hash;
    DWORD Reserved2;
} PREFETCH_FILE_HEADER;

#pragma pack(pop)
```

Now, you can do something like this:
```
PREFETCH_COMPRESSED_HEADER * pCompressedHeader = (... your data...)
PREFETCH_FILE_HEADER * pHeader;
if (pCompressedHeader->Signature == '\X04MAM') { // You can do something fancy like this for "MAM\x04"
    // call your decompression here
   pHeader = decompressed data // set pHeader to the beginning of the decompressed data.
} else {
    pHeader = pCompressedHeader; // I am being liberal here with type-casting
}

```

Now, at `pHeader->FileName`, you have the first 29 characters of your filename, and you're set.

What are your thoughts?

Reply:
is there an advantage of using structs over the current implementation (ex: safety, huge speed increases, portability)?

Reply:
The primary benefit in this case is to simplify and improve readability.  Information is now available as fields in the structures, and there is no need to perform individual casts, e.g., `*reinterpret_cast<int*>(&prefetch_data[12])`. Constructing a `string_view` is no longer necessary nor is the function `ucharToString`, since you can now use osquery's existing conversions, such as `wstringToString(pHeader->FileName)` to retrieve the file name in utf-16. There is a speed benefit to doing this, but I consider this secondary.

Sorry for the late replies and not getting a chance to review these changes sooner. I've been juggling with a couple of other things in real-life.

Reply:
so i switched to wstring_view instead of string_view for the filename and accessed data parsing and use `wstirngToString` to convert it to a string.  It does use `reinterpret_cast` but that should be ok since the filename and accessed_data are UTF16? Also removed `ucharToString` since its now using `wstringToString`

Reply:
> so i switched to wstring_view instead of string_view for the filename and accessed data parsing and use `wstirngToString` to convert it to a string.
> It does use `reinterpret_cast` but that should be ok since the filename and accessed_data are UTF16?

Unfortunately not. It's still violating the strict aliasing rule in C++, because you will dereference the data (the `char` buffer) through a pointer that cannot alias it (`wchar_t *`). It cannot alias it because it's not compatible.
It's not compatible because C++ reasons with an object model, there's no `wchar_t` being constructed there or previously constructed; it sees only an array of characters.

Reply:
ok, thats good to know. Reverted back to previous implementation which should still work fine?  
I tried researching other conversion functions but did not have any luck, they all require additional libraries.  
The `ucharToString` function should work since its skipping the every other char (`00`) until it reaches the null terminator.  
ex:`37 00 5A 00 46 00 4D 00 2E 00 45 00 58 00 45 00 00 00`. 
 skips every other char until it reaches null terminator

Reply:
The functions I was referring to earlier do the conversion already (wstringToString), but one need to add an overload to accept a char  * and a size; also the string needs to be converted to UTF8 (which the above function does).

I do not suggest to roll your own; the one you've written has a bug in strnlen (the max size needs to be reduced with the remaining characters from the point you're applying the `offset` to the end of the string).

Reply:
@Smjert im looking at the functions in `wstringToString`:
- WideCharToMultiByte
- MultiByteToWideChar

None of them seem to accept `char` UTF16 values?
- WideCharToMultiByte expects a wchar
- MultiByteToWideChar doesnt appear to support UTF16 encoding as input

Is there another windows function that needs to be used, or is a there a way to convert the data to acceptable format?

Reply:
wchar_t is UTF-16 on Windows. Considering the data as a struct I discussed previously (following libscca documentation) would simplify things, as the truncated file name would be available as a field in the struct as a WCHAR array.

```
#pragma pack(push, 1)
typedef struct _PREFETCH_COMPRESSED_HEADER {
    DWORD Signature;
    DWORD TotalUncompressedSize;
    BYTE CompressedData[1]; // arbitrary size
} PREFETCH_COMPRESSED_HEADER;

typedef struct _PREFETCH_FILE_HEADER {
    DWORD Version;
    DWORD Signature;
    DWORD Reserved1;
    DWORD FileSize;
    WCHAR FileName[30];
    DWORD Hash;
    DWORD Reserved2;
} PREFETCH_FILE_HEADER;

#pragma pack(pop)
```

Reply:
@farfella  i looked in to that but that only solves part of the problem?  
The UTF16 issue affects 3 parts:
- filename (60 bytes)
- accessed files (size is dynamic, each entry ends in null char)
- accessed directories (size is dynamic, each entry ends in null char)

The struct solution can fix the filename UTF16 issue, but i dont think it can fix the access files or directories issue?  
Since the UTF16 issue is the same for all three, im kind of hoping there is solution that can be used for all three?

Reply:
I will take a looksie this weekend. Apologies for the delay-- real life things.

Reply:
ok thanks!, in the mean time since the current implementation has a bug I switched to the original implementation (looping through the vector, until it encounters a null terminator). I added a few additions to address the original comments on the implementation